pallets/energy/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+//! # Energy Module
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+    use frame_support::traits::Currency;
+    use pallet_transaction_payment::OnChargeTransaction;
+    use sp_runtime::ArithmeticError;
+    use sp_runtime::traits::{CheckedAdd, CheckedSub, DispatchInfoOf, PostDispatchInfoOf, Saturating, StaticLookup, Zero};
+    use sp_std::convert::TryInto;
+    use crate::*;
+
+    pub(crate) type BalanceOf<T> =
+    <<T as Config>::Currency as Currency<<T as frame_system::pallet::Config>::AccountId>>::Balance;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_transaction_payment::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// The currency type.
+        type Currency: Currency<Self::AccountId>;
+
+        /// The fallback [OnChargeTransaction] that should be used if there is not enough energy to
+        /// pay the transaction fees.
+        type FallbackOnChargeTransaction: OnChargeTransaction<Self, Balance=BalanceOf<Self>>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub (super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Energy have been generated to an account.
+        EnergyGenerated {
+            /// The account that generated the energy.
+            generator: T::AccountId,
+            /// The account that received the energy.
+            receiver: T::AccountId,
+            /// The amount of balance that was burnt.
+            burnt_balance: BalanceOf<T>,
+            /// The amount of energy that have been generated.
+            generated_energy: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Not enough balance to generate energy.
+        NotEnoughBalance,
+    }
+
+    /// Total energy generated.
+    #[pallet::storage]
+    #[pallet::getter(fn total_energy)]
+    pub(crate) type TotalEnergy<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
```

Comment:
How are we using this storage?
Also, do we want to have `BurnedEnergy` storage value?

Reply:
It tracks the total issued energy in the system, like Balances pallet.

I don't if we need `BurnedEnergy`, but adding it will be easy

Reply:
> It tracks the total issued energy in the system, like Balances pallet.
> 
> I don't know if we need `BurnedEnergy`, but adding it will be easy

I think we don't need to add it. It may be useful in future to have `TotalEnergy`, and it's reasonable to add it now.
But I don't see a use-case where we'd need `BurnedEnergy`. To just get a value we can index our chain using Subsquid or something.